x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import dateMath from '@elastic/datemath';
+import { i18n } from '@kbn/i18n';
+
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+import chrome from 'ui/chrome';
+import { API_BASE_URL_V1 } from '../../common/constants';
+
+const API_BASE_URL = `${API_BASE_URL_V1}/generate/immediate/csv/saved-object`;
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('xpack.reporting.dashboard.downloadCsvPanelTitle', {
+          defaultMessage: 'Download CSV',
+        }),
+        id: 'downloadCsvReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    const adapters = searchEmbeddable.getInspectorAdapters();
+    if (!adapters) {
+      return {};
+    }
+
+    if (adapters.requests.requests.length === 0) {
+      return {};
+    }
+
+    return searchEmbeddable.searchScope.searchSource.getSearchRequestBody();
+  }
+
+  // @TODO: Clean this up once we update SavedSearch's interface
+  // and location in the file-system. `viewMode` also has an enum
+  // buried inside of the dashboard folder we could use vs a bare string
+  public isVisible = (panelActionAPI: PanelActionAPI): boolean => {
+    const { embeddable, containerState } = panelActionAPI;
+
+    return (
+      containerState.viewMode !== 'edit' && !!embeddable && embeddable.hasOwnProperty('savedSearch')
+    );
+  };
+
+  public onClick = async (panelActionAPI: PanelActionAPI) => {
+    const { embeddable } = panelActionAPI as any;
+    const {
+      timeRange: { from, to },
+    } = embeddable;
+
+    if (!embeddable) {
+      return;
+    }
+
+    const searchEmbeddable = embeddable;
+    const state = await this.generateJobParams({ searchEmbeddable });
+
+    const id = `search:${embeddable.savedSearch.id}`;
+    const filename = embeddable.getPanelTitle();
+    const timezone = chrome.getUiSettingsClient().get('dateFormat:tz');
+    const fromTime = dateMath.parse(from);
+    const toTime = dateMath.parse(to);
+
+    if (!fromTime || !toTime) {
+      return this.onGenerationFail(
+        new Error(`Invalid time range: From: ${fromTime}, To: ${toTime}`)
+      );
+    }
+
+    const body = JSON.stringify({
+      timerange: {
```

Comment:
In hindsight, it seems like an overcomplication to support an additional field for timerange in the POST payload.

You get a search body from `searchEmbeddable.searchScope.searchSource.getSearchRequestBody()`, and that has the same filter information from the time picker that you're additionally getting via `panelActionAPI.embeddable`

It would be a medium-ish change to take out this field, but I think it would be worth it as it would simplify the composition of different filters, and not duplicate a time filter.

See https://github.com/elastic/kibana/pull/34571/files#diff-6ac014de80a63efc019c68e55e498b9bR25 , lines 25-52 on where this timefilter is combined with `searchSourceFilter`, effectively making the query have 2 of the same time filter.

We should just keep the `searchSourceFilter` information. That simplification would also get rid of this unruly business: https://github.com/elastic/kibana/pull/34571/files#diff-6ac014de80a63efc019c68e55e498b9bR38